pkg/controller/config-controller.go

```diff
@@ -1,377 +1,190 @@
-	}
-	klog.V(3).Infof("Processing object: %s", object.GetName())
+// Reconcile the reconcile loop for the CDIConfig object.
+func (r *CDIConfigReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.Log.WithValues("CDIConfig", req.NamespacedName)
+	log.Info("reconciling CDIConfig")
 
-	var config *cdiv1.CDIConfig
-	config, err := c.configLister.Get(c.configName)
-	if err != nil {
-		runtime.HandleError(errors.Errorf("error getting CDI config: %s", err))
-		return
+	var config cdiv1.CDIConfig = *MakeEmptyCDIConfigSpec(r.ConfigName)
+	if err := SetOwnerToOperator(&r.Client, &config, r.CDINamespace); err != nil {
+		return reconcile.Result{}, err
 	}
-
-	if ing, ok := obj.(*extensionsv1beta1.Ingress); ok {
-		for _, rule := range ing.Spec.Rules {
-			for _, path := range rule.HTTP.Paths {
-				if path.Backend.ServiceName == c.uploadProxyServiceName {
-					c.enqueueCDIConfig(config)
-					return
-				}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: r.ConfigName}, &config); err != nil {
+		if errors.IsNotFound(err) {
+			if err := r.Client.Create(context.TODO(), &config); err != nil {
+				// If the object already exists, just error out, and next reconcile the Get will return a value.
+				return reconcile.Result{}, err
 			}
+		} else {
+			log.Error(err, "Unable to get CDIConfig object")
+			return reconcile.Result{}, err
 		}
 	}
-	if route, ok := obj.(*routev1.Route); ok {
-		if route.Spec.To.Name == c.uploadProxyServiceName {
-			c.enqueueCDIConfig(config)
-			return
+	// Keep a copy of the original for comparison later.
+	currentConfigCopy := config.DeepCopyObject()
+
+	config.Status.UploadProxyURL = config.Spec.UploadProxyURLOverride
+	// No override, try Ingress
+	if config.Status.UploadProxyURL == nil {
+		if err := r.reconcileIngress(&config); err != nil {
+			log.Error(err, "Unable to reconcile Ingress")
+			return reconcile.Result{}, err
 		}
 	}
-	if conf, ok := obj.(*cdiv1.CDIConfig); ok {
-		if conf.Name == config.Name {
-			c.enqueueCDIConfig(conf)
-			return
+	// No override or Ingress, try Route
+	if config.Status.UploadProxyURL == nil {
+		if err := r.reconcileRoute(&config); err != nil {
+			log.Error(err, "Unable to reconcile Routes")
+			return reconcile.Result{}, err
 		}
 	}
 
-	return
-}
-
-func (c *ConfigController) runWorker() {
-	for c.processNextWorkItem() {
-	}
-}
-
-func (c *ConfigController) processNextWorkItem() bool {
-	obj, shutdown := c.queue.Get()
-
-	if shutdown {
-		return false
+	if err := r.reconcileStorageClass(&config); err != nil {
+		return reconcile.Result{}, err
 	}
 
-	err := func(obj interface{}) error {
-		defer c.queue.Done(obj)
-
-		var key string
-		var ok bool
-
-		if key, ok = obj.(string); !ok {
-			c.queue.Forget(obj)
-			runtime.HandleError(errors.Errorf("expected string in workqueue but got %#v", obj))
-			return nil
+	if !reflect.DeepEqual(currentConfigCopy, config) {
```

Comment:
as is `currentConfigCopy` is interface with a pointer receiver.  config is a struct (not pointer).  Does `DeepEqual` do the right thing given that?

Reply:
Good question, need to double check.

Reply:
Since I am making config a pointer, then deep equal will do the right thing.

Reply:
Types should match now, so deep equal will do the right thing.